src/HDF5.jl

```diff
@@ -1370,12 +1378,56 @@ Base.getindex(parent::Union{File,Group}, r::Reference) = _deref(parent, r)
 Base.getindex(parent::Dataset, r::Reference) = _deref(parent, r) # defined separately to resolve ambiguity
 
 # convert special types to native julia types
-normalize_types(x) = x
-normalize_types(x::NamedTuple{T}) where {T} = NamedTuple{T}(map(normalize_types, values(x)))
-normalize_types(x::Cstring) = unsafe_string(x)
-normalize_types(x::FixedString) = unpad(String(collect(x.data)), pad(x))
-normalize_types(x::FixedArray) = normalize_types.(reshape(collect(x.data), size(x)...))
-normalize_types(x::VariableArray) = normalize_types.(copy(unsafe_wrap(Array, convert(Ptr{eltype(x)}, x.p), x.len, own=false)))
+function normalize_types(::Type{T}, buf::AbstractMatrix{UInt8}) where {T}
+    return [_normalize_types(T, view(buf, :, ind)) for ind in axes(buf, 2)]
+end
+
+# high-level description which should always work --- here, the buffer contains the bytes
+# for exactly 1 element of an object of type T, so reinterpret the `UInt8` vector as a
+# length-1 array of type `T` and extract the (only) element.
+function _typed_load(::Type{T}, buf::AbstractVector{UInt8}) where {T}
+    return @inbounds reinterpret(T, buf)[1]
+end
+# fast-path for common concrete types with simple layout (which should be nearly all
+# cases)
+function _typed_load(::Type{T}, buf::V) where {T, V <: Union{Vector{UInt8}, Base.FastContiguousSubArray{UInt8,1}}}
+    dest = Ref{T}()
+    GC.@preserve buf begin
+        srcptr  = pointer(buf)
+        destptr = unsafe_convert(Ptr{Cvoid}, dest)
+        Base._memcpy!(destptr, srcptr, sizeof(T))
+    end
+    return dest[]
+end
+
+_normalize_types(::Type{T}, buf::AbstractVector{UInt8}) where {T} = _typed_load(T, buf)
+function _normalize_types(::Type{T}, buf::AbstractVector{UInt8}) where {K, T <: NamedTuple{K}}
+    nv = ntuple(length(K)) do ii
+        elT = fieldtype(T, ii)
+        off = fieldoffset(T, ii) % Int
+        sub = view(buf, off .+ (1:sizeof(elT)))
+        return _normalize_types(elT, sub)
+    end
+    return NamedTuple{K}(nv)
+end
+function _normalize_types(::Type{V}, buf::AbstractVector{UInt8}) where {T, V <: VariableArray{T}}
+    va = _typed_load(V, buf)
+    pbuf = unsafe_wrap(Array, convert(Ptr{UInt8}, va.p), (sizeof(T), Int(va.len)), own = false)
+    if do_normalize(T)
+        return normalize_types(T, pbuf)
+    else
+        return copy(vec(reinterpret(T, pbuf)))
+    end
+end
+function _normalize_types(::Type{F}, buf::AbstractVector{UInt8}) where {T, F <: FixedArray{T}}
+    if do_normalize(T)
+        return reshape(normalize_types(T, reshape(buf, sizeof(T), :)), size(F)...)
+    else
+        return copy(reshape(reinterpret(T, buf), size(F)...))
```

Comment:
Is this `copy` defensive? Doesn't the reshape in this instance  copy the buffer anyways? Same question in the directly above method.

Reply:
The `copy` is to (1) make the returned array independent of the original, parent buffer in `read()` since some of the wrapper types will just keep working with a view into the original buffer, and (2) it gets rid of possibly complex nested types:

For instance:
```julia
buf = reshape(collect(UInt8(1):UInt8(64)), 32, 2) # in read()
compound = view(buf, :, 1)  # in normalize_types()
pair = view(compound, 1:16) # in _normalize_types(::Type{NamedTuple})
el = reshape(reinterpret(UInt32, pair), (2, 2)) # in _normalize_types(::Type{FixedArray})
```
```julia
julia> typeof(el)
Base.ReshapedArray{UInt32, 2, Base.ReinterpretArray{UInt32, 1, UInt8, SubArray{UInt8, 1, Matrix{UInt8}, Tuple{UnitRange{Int64}, Int64}, true}, false}, Tuple{}}

julia> typeof(copy(el))
Matrix{UInt32} (alias for Array{UInt32, 2})
```
After potentially arbitrary levels of nesting, I think `copy`ing is the only way to guarantee you don't have arbitrarily complex types being returned.